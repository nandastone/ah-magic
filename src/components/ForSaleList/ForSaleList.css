.c-ForSaleList {

  font-size: 14px;
  table-layout: fixed;
}

  .c-ForSaleList th { border-top: none; }
    .c-ForSaleList th .sort-icon { margin-left: 0.5em; }

/* Elements */

.c-ForSaleList__row td { vertical-align: middle; }

/* Modifiers */
