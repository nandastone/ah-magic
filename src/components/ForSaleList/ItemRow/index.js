import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import moment from 'moment'

import { calculateListingFinances } from '../../../utils'

// Components

import { Badge, Button, ButtonGroup } from 'reactstrap'
import ItemDetailModal from '../../ItemDetailModal'
import ItemTitle from '../../ItemTitle'
import ListItemModal from '../../ListItemModal'
import SellItemModal from '../../SellItemModal'
import WowCurrency from '../../WowCurrency'

class ItemRow extends PureComponent {
  state = {
    isDetailsExpanded: false,
    isListExpanded: false,
    isSoldExpanded: false
  }

  // Event handling

  handleClickRow = (event) => {
    event.preventDefault()
    this.setState({ isDetailsExpanded: true })
  }

  handleClickList = (event) => {
    event.stopPropagation()
    this.setState({ isListExpanded: true })
  }

  handleClickEnd = (event) => {
    event.stopPropagation()
    this.props.onEnd()
  }

  handleClickSold = (event) => {
    event.stopPropagation()
    this.setState({ isSoldExpanded: true })
  }

  handleClickDelete = (event) => {
    event.stopPropagation()
    if (window.confirm('Are you sure want to delete this item?')) {
      this.props.onDelete()
    }
  }

  handleCompleteList = (payload) => {
    this.props.onList(payload)
  }

  handleCompleteSold = (payload) => {
    this.props.onSold(payload)
  }

  handleCloseModal = () => {
    this.setState({
      isDetailsExpanded: false,
      isListExpanded: false,
      isSoldExpanded: false
    })
  }

  // Rendering

  render () {
    const firstHistory = _.first(this.props.item.history)
    const lastHistory = _.last(this.props.item.history)
    const isListed = lastHistory && lastHistory.type === 'listing' && !lastHistory.endedAt
    const { cost, price, profit } = calculateListingFinances(this.props.item.history)
    const createdAt = firstHistory ? firstHistory.createdAt : null
    const updatedAt = lastHistory ? (lastHistory.endedAt || lastHistory.createdAt) : null

    return (
      <tr
        className={'c-ForSaleList__row'}
        onClick={this.handleClickRow}
      >
        <td>
          <React.Fragment>
            <ItemTitle name={this.props.item.name} count={this.props.item.stackable} />
            <ItemDetailModal
              item={this.props.item}
              open={this.state.isDetailsExpanded}
              onClose={this.handleCloseModal}
            />
          </React.Fragment>
        </td>
        <td>
          <WowCurrency value={cost} />
        </td>
        <td>
          <WowCurrency value={price} />
        </td>
        <td>
          <Badge color={profit > 0 ? 'success' : 'danger'}>
            {profit <= 0 ? '-' : ''}
            <WowCurrency value={Math.abs(profit)} />
          </Badge>
        </td>
        <td>
          {
            createdAt
            ? <span title={moment(createdAt).format('D MMM YYYY, h:mm:ss a')} data-test='created-at'>
                {moment(createdAt).format('D MMM YYYY')}
              </span>
            : null
          }
        </td>
        <td>
          {
            updatedAt
            ? <span title={moment(updatedAt).format('D MMM YYYY, h:mm:ss a')} data-test='updated-at'>
                {moment(updatedAt).format('D MMM YYYY')}
              </span>
            : null
          }
        </td>
        <td>
          <ButtonGroup size='sm'>
            {
              !isListed
              ? <Button
                  color='secondary'
                  onClick={this.handleClickList}
                  data-test='list-button'
                >
                  List
                </Button>
              : <Button
                  color='secondary'
                  onClick={this.handleClickEnd}
                  data-test='end-button'
                >
                  End
                </Button>
            }
            <Button
              color='primary'
              onClick={this.handleClickSold}
              data-test='sold-button'
            >
              Sold
            </Button>
            <Button
              color='danger'
              onClick={this.handleClickDelete}
              data-test='delete-button'
            >
              Delete
            </Button>
          </ButtonGroup>
          <ListItemModal
            item={this.props.item}
            open={this.state.isListExpanded}
            onComplete={this.handleCompleteList}
            onClose={this.handleCloseModal}
          />
          <SellItemModal
            item={this.props.item}
            open={this.state.isSoldExpanded}
            onComplete={this.handleCompleteSold}
            onClose={this.handleCloseModal}
          />
        </td>
      </tr>
    )
  }
}

ItemRow.defaultProps = {
  onList: () => {},
  onSold: () => {},
  onEnd: () => {},
  onDelete: () => {}
}

ItemRow.propTypes = {
  item: PropTypes.object.isRequired,
  onList: PropTypes.func,
  onSold: PropTypes.func,
  onEnd: PropTypes.func,
  onDelete: PropTypes.func
}

export default ItemRow
