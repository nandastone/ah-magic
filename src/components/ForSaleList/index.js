import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import uuid from 'uuid/v4'
import moment from 'moment'

import {
  calculateAHListingCost,
  calculateAHTransactionCost,
  calculateItemFinances
} from '../../utils'

// Components

import { Table } from 'reactstrap'
import ItemRow from './ItemRow'

// Assets

import './ForSaleList.css'

class ForSaleList extends PureComponent {
  state = {
    sortField: '',
    sortDirection: 'DESC'
  }

  // Event handling

  handleListItem = (item, { bid, price, duration }) => {
    const updatedItem = {
      ...item,
      history: [
        ...item.history,
        {
          key: uuid(),
          type: 'listing',
          duration,
          bid,
          price,
          cost: calculateAHTransactionCost(price),
          createdAt: moment().format()
        }
      ]
    }
    this.props.onChangeItem(updatedItem)
  }

  handleSoldItem = (item, { price, saleType }) => {
    const lastHistory = _.last(item.history)
    let updatedItem = { ...item }

    // If the item is listed, remove the "listing" history entry to later replace with a "sale" history entry.
    if (lastHistory && lastHistory.type === 'listing') {
      updatedItem = {
        ...updatedItem,
        history: [
          ..._.dropRight(item.history),
          {
            ...lastHistory,
            // If we sold via the AH, negate the previous listing fee.
            cost: saleType === 'ah' ? 0 : lastHistory.cost,
            // Save the listing end date.
            endedAt: lastHistory.endedAt || moment().format()
          }
        ]
      }
    }

    // Create a "sale" history entry.
    updatedItem = {
      ...updatedItem,
      history: [
        ...updatedItem.history,
        {
          key: uuid(),
          type: 'sale',
          saleType,
          price,
          // Only add transaction fees if the item was sold on the AH.
          cost: saleType === 'ah' ? calculateAHTransactionCost(price) : 0,
          createdAt: moment().format(),
          endedAt: moment().format()
        }
      ]
    }

    this.props.onChangeItem(updatedItem)
  }

  handleEndItem = (item) => {
    const lastHistory = _.last(item.history)
    const updatedItem = {
      ...item,
      // Update last history item (listing) with an ended date
      history: [
        ..._.dropRight(item.history),
        {
          ...lastHistory,
          cost: calculateAHListingCost(item.vendorValue, lastHistory.duration),
          endedAt: moment().format()
        }
      ]
    }
    this.props.onChangeItem(updatedItem)
  }

  handleDeleteItem = (item) => {
    this.props.onDeleteItem(item)
  }

  // Private

  _sort (field) {
    if (this.state.sortField === field && this.state.sortDirection === 'ASC') {
      // Reset sorting if we've reached the end of sort directions.
      this.setState({ sortField: '', sortDirection: 'DESC' })
    } else if (this.state.sortField === field) {
      this.setState({ sortField: field, sortDirection: 'ASC' })
    } else {
      this.setState({ sortField: field, sortDirection: 'DESC' })
    }
  }

  _getSortedItems (items) {
    const { sortField: field, sortDirection: direction } = this.state
    let sorted = [...items]

    switch (field) {
      case 'name':
        sorted = _.sortBy(items, ['name'])
        if (direction === 'DESC') sorted.reverse()
        break
      case 'cost':
        sorted = _.sortBy(items, (item) => {
          const { cost } = calculateItemFinances(item.history)
          return cost || 0
        })
        if (direction === 'DESC') sorted.reverse()
        break
      case 'price':
        sorted = _.sortBy(items, (item) => {
          const { price } = calculateItemFinances(item.history)
          return price || 0
        })
        if (direction === 'DESC') sorted.reverse()
        break
      case 'profit':
        sorted = _.sortBy(items, (item) => {
          const { profit } = calculateItemFinances(item.history)
          return profit || 0
        })
        if (direction === 'DESC') sorted.reverse()
        break
      case 'createdAt':
        sorted = _.sortBy(items, (item) => {
          const firstHistory = _.first(item.history)
          return firstHistory.createdAt ? moment(firstHistory.createdAt).format('X') : 0
        })
        if (direction === 'DESC') sorted.reverse()
        break
      case 'updatedAt':
        sorted = _.sortBy(items, (item) => {
          const lastHistory = _.last(item.history)
          const updatedAt = (lastHistory.endedAt || lastHistory.createdAt)
          return updatedAt ? moment(updatedAt).format('X') : 0
        })
        if (direction === 'DESC') sorted.reverse()
        break
      default:
        throw new Error(`Invalid sorting column passed to ForSaleList: ${field}`)
    }

    return sorted
  }

  // Rendering

  render () {
    const sortedItems = this.state.sortField ? this._getSortedItems(this.props.items) : this.props.items

    return (
      <Table hover className='c-ForSaleList'>
        <thead>
          <tr>
            <th
              onClick={() => this._sort('name')}
              className='w-25'
            >
              Name
              {
                this.state.sortField === 'name'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            <th
              onClick={() => this._sort('cost')}
              className='w-10'
            >
              Cost
              {
                this.state.sortField === 'cost'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            <th
              onClick={() => this._sort('price')}
              className='w-10'
            >
              Price
              {
                this.state.sortField === 'price'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            <th
              onClick={() => this._sort('profit')}
              className='w-10'
            >
              Profit
              {
                this.state.sortField === 'profit'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            <th
              onClick={() => this._sort('createdAt')}
              className='w-15'
            >
              Created
              {
                this.state.sortField === 'createdAt'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            <th
              onClick={() => this._sort('updatedAt')}
              className='w-15'
            >
              Updated
              {
                this.state.sortField === 'updatedAt'
                ? ` Sorted (${this.state.sortDirection})`
                : null
              }
            </th>
            {/* Actions column */}
            <th className='w-15'></th>
          </tr>
        </thead>
        <tbody>
          {sortedItems.map((item) => {
            return (
              <ItemRow
                key={item.key}
                item={item}
                onList={(payload) => this.handleListItem(item, payload)}
                onSold={(payload) => this.handleSoldItem(item, payload)}
                onEnd={() => this.handleEndItem(item)}
                onDelete={() => this.handleDeleteItem(item)}
              />
            )
          })}
        </tbody>
      </Table>
    )
  }
}

ForSaleList.defaultProps = {
  onChangeItem: () => {},
  onDeleteItem: () => {}
}

ForSaleList.propTypes = {
  items: PropTypes.array,
  onChangeItem: PropTypes.func,
  onDeleteItem: PropTypes.func,
}

export default ForSaleList
