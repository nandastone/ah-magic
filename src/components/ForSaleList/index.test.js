import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

// Components

import ForSaleList from './'

const items = [
  { key: 1, name: 'Item 1', history: [], vendorValue: 5000 },
  { key: 2, name: 'Item 2', history: [], vendorValue: 10000 },
  { key: 3, name: 'Item 3', history: [], vendorValue: 10000 },
]

describe('ForSaleList component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<ForSaleList items={items} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render the correct number of item rows', () => {
    const wrapper = shallow(<ForSaleList items={items} />)
    expect(wrapper.find('ItemRow').length).toEqual(3)
  })
})
