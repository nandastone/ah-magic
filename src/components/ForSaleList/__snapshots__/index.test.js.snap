// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForSaleList component renders correctly 1`] = `
<table
  className="c-ForSaleList table table-hover"
>
  <thead>
    <tr>
      <th
        className="w-25"
        onClick={[Function]}
      >
        Name
      </th>
      <th
        className="w-10"
        onClick={[Function]}
      >
        Cost
      </th>
      <th
        className="w-10"
        onClick={[Function]}
      >
        Price
      </th>
      <th
        className="w-10"
        onClick={[Function]}
      >
        Profit
      </th>
      <th
        className="w-15"
        onClick={[Function]}
      >
        Created
      </th>
      <th
        className="w-15"
        onClick={[Function]}
      >
        Updated
      </th>
      <th
        className="w-15"
      />
    </tr>
  </thead>
  <tbody>
    <tr
      className="c-ForSaleList__row"
      onClick={[Function]}
    >
      <td>
        <span
          data-test="name"
        >
          Item 1
        </span>
         
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="badge badge-danger"
        >
          -
          <span
            className="c-WowCurrency"
          >
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
            >
              0
              <img
                alt="Gold"
                src="gold.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
            >
              0
              <img
                alt="Silver"
                src="silver.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
            >
              0
              <img
                alt="Copper"
                src="copper.png"
              />
            </span>
          </span>
        </span>
      </td>
      <td />
      <td />
      <td>
        <div
          className="btn-group-sm btn-group"
          role="group"
        >
          <button
            className="btn btn-secondary"
            data-test="list-button"
            onClick={[Function]}
            type="button"
          >
            List
          </button>
          <button
            className="btn btn-primary"
            data-test="sold-button"
            onClick={[Function]}
            type="button"
          >
            Sold
          </button>
          <button
            className="btn btn-danger"
            data-test="delete-button"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </div>
      </td>
    </tr>
    <tr
      className="c-ForSaleList__row"
      onClick={[Function]}
    >
      <td>
        <span
          data-test="name"
        >
          Item 2
        </span>
         
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="badge badge-danger"
        >
          -
          <span
            className="c-WowCurrency"
          >
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
            >
              0
              <img
                alt="Gold"
                src="gold.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
            >
              0
              <img
                alt="Silver"
                src="silver.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
            >
              0
              <img
                alt="Copper"
                src="copper.png"
              />
            </span>
          </span>
        </span>
      </td>
      <td />
      <td />
      <td>
        <div
          className="btn-group-sm btn-group"
          role="group"
        >
          <button
            className="btn btn-secondary"
            data-test="list-button"
            onClick={[Function]}
            type="button"
          >
            List
          </button>
          <button
            className="btn btn-primary"
            data-test="sold-button"
            onClick={[Function]}
            type="button"
          >
            Sold
          </button>
          <button
            className="btn btn-danger"
            data-test="delete-button"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </div>
      </td>
    </tr>
    <tr
      className="c-ForSaleList__row"
      onClick={[Function]}
    >
      <td>
        <span
          data-test="name"
        >
          Item 3
        </span>
         
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="c-WowCurrency"
        >
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
          >
            0
            <img
              alt="Gold"
              src="gold.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
          >
            0
            <img
              alt="Silver"
              src="silver.png"
            />
          </span>
          <span
            className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
          >
            0
            <img
              alt="Copper"
              src="copper.png"
            />
          </span>
        </span>
      </td>
      <td>
        <span
          className="badge badge-danger"
        >
          -
          <span
            className="c-WowCurrency"
          >
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--gold"
            >
              0
              <img
                alt="Gold"
                src="gold.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--silver"
            >
              0
              <img
                alt="Silver"
                src="silver.png"
              />
            </span>
            <span
              className="c-WowCurrency__coin c-WowCurrenct__coin--copper"
            >
              0
              <img
                alt="Copper"
                src="copper.png"
              />
            </span>
          </span>
        </span>
      </td>
      <td />
      <td />
      <td>
        <div
          className="btn-group-sm btn-group"
          role="group"
        >
          <button
            className="btn btn-secondary"
            data-test="list-button"
            onClick={[Function]}
            type="button"
          >
            List
          </button>
          <button
            className="btn btn-primary"
            data-test="sold-button"
            onClick={[Function]}
            type="button"
          >
            Sold
          </button>
          <button
            className="btn btn-danger"
            data-test="delete-button"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
